### ISSUE TITLE
experiment: run first measured Granite 4.1 base-vs-SFT intervention

### ISSUE THREAD

**Issue by rmems** [rmems/agoge-forger#101]
## Why this is reopened

This issue was closed by temporary debt-mode triage, not because the Granite experiment ran. It is now restored as the **single measured flagship experiment** for Agoge.

The old G0/G1/G2/reference matrix is intentionally reduced. This issue does one causal comparison only:

```text
G0 — untouched ibm-granite/granite-4.1-3b-base
G1 — the same pinned base + one frozen code-repair SFT intervention
```

Mixtures, public-SWE arms, Prometheus composition tests, and external reference checkpoints remain parked until this result is complete and publishable.

## Research question

> Under one frozen code-repair dataset, training objective, and held-out evaluation contract, does a bounded SFT intervention change Granite 4.1 3B Base performance relative to the exact untouched starting checkpoint?

Positive, negative, null, mixed, or inconclusive are all valid outcomes.

## Blockers

Do not begin measured G0/G1 execution until:

- #133 proves completion-only loss and completion-preserving truncation on the real trainer batch;
- #104 qualifies MiniCPM5 and then Granite 4.1 through load → train → save → clean reload → generation;
- #100 can execute and persist the paired held-out evaluation;
- the selected training dataset/split is pinned by immutable revision/digest.

## Freeze before measured G0

Record before inspecting the measured baseline:

- exact Granite model revision;
- exact tokenizer revision;
- exact Agoge commit;
- frozen train/validation/held-out split digest;
- serializer/rendering implementation + hash/version;
- prompt/completion loss-masking policy;
- context length and truncation policy;
- LoRA/QLoRA target modules and quantization config;
- optimizer/LR, effective batch size, update/epoch budget, seed;
- decoding/scoring contract and primary held-out metric;
- protected regression/no-degradation metric if applicable;
- interpretation rule for invalid/inconclusive cases.

Do not change these after seeing G0/G1 results without versioning a new experiment.

## Execution

1. Freeze and commit the experiment contract.
2. Run G0 on the untouched held-out set.
3. Execute the bounded G1 SFT run.
4. Preserve checkpoints/final adapter, tokenizer/config, run manifest, telemetry, and artifact hashes.
5. Clean-reload G1.
6. Evaluate G1 under the identical logical held-out contract.
7. Produce machine-readable paired deltas and a concise report.

## Required measurements

- held-out task metric(s), G0 and G1;
- G0 → G1 delta;
- regressions, new successes, ties, invalid/inconclusive cases;
- training/validation loss as diagnostic data only;
- accepted training tokens / record count;
- wall-clock training time and throughput;
- peak VRAM;
- checkpoint/adapter sizes and hashes;
- exact software/CUDA/GPU environment.

## Required artifacts

```text
reports/granite-4.1-first-sft/<experiment_id>/
  experiment-contract.json
  g0-base/
  g1-sft/
  comparison.json
  regressions.jsonl
  training-results.json
  report.md
  reproduction.md
```

## Acceptance criteria

- [ ] Every blocker above is satisfied before measured execution.
- [ ] G0 is evaluated before G1 training begins.
- [ ] G1 completes or leaves a preserved, diagnosable failure artifact.
- [ ] G0 and G1 use the same held-out task IDs, rendering contract, decoding settings, and scoring version.
- [ ] Exact model/tokenizer/data/code revisions and artifact hashes are recorded.
- [ ] The report states the measured result without changing success criteria post hoc.
- [ ] Reproduction commands are committed.

## Non-goals

- G2 Synthetic + Prometheus mixture;
- public-SWE A–F ablation (#103 remains parked);
- Granite 4.2 or other reference benchmarking;
- Mellum/LFM/HRM replication;
- DPO/RL;
- multi-GPU or cloud training;
- hyperparameter search after seeing the baseline.

## Downstream

After this issue completes, reopen #102 to publish the first SFT artifact/reproducibility bundle. Only after #102 should broader ablations or later post-training methods become active work.

**Comment by rmems** [comment]
**DELETE** — 04 Agoge Technical Debt Reducer

Debt-mode KEEP/DELETE triage for `rmems/agoge-forger#101`.
This is a training/experiment/roadmap item. Agoge maintenance mode forbids new training mills, model downloads, eval campaigns, and cloud jobs. Recommend close unless Raul reopens with a named train brief.
